File: miscellaneous/ltcd_2420.cpp
//LeetCode - problem 2420

    vector<int> goodIndices(vector<int>& nums, int k) {
        int n = nums.size();
        
        
        vector<int> ans;
        
        vector<int> front(n, 0); 
        vector<int> back(n, 0);
        
        
        front[0] = 0;
        front[1] = 1;
        for(int i = 2; i < n; i ++) {
            if(nums[i - 1] <= nums[i - 2]) {
                front[i] = 1 + front[i - 1];
            } else {
                front[i] = 1;
            }
        }
        
        back[n - 1] = 0;
        back[n - 2] = 1;
        for(int i = n - 3; i >= 0; i --) {
            if(nums[i + 1] <= nums[i + 2]) {
                back[i] = 1 + back[i + 1];
            } else {
                back[i] = 1;
            }
        }
        
        for(int i = k; i < n - k; i ++) {
            if(min(front[i], back[i]) >= k) ans.push_back(i);   
        }
        
        return ans;
    }

File: miscellaneous/ltcd_2420.py
'''
LeetCode - problem 2420
'''

    def goodIndices(self, nums: List[int], k: int) -> List[int]:
        ans = []
        non_increasing_dp = [1] * len(nums)
        non_decreasing_dp = [1] * len(nums)
        def check_non_increasing():
            for i in range(1, len(nums)):
                if nums[i] <= nums[i - 1]:
                    non_increasing_dp[i] = non_increasing_dp[i - 1] + 1
            
        def check_non_decreasing():
            for i in range(len(nums) - 2, 0, -1):
                if nums[i] <= nums[i + 1]:
                    non_decreasing_dp[i] = non_decreasing_dp[i + 1] + 1
          
        check_non_increasing()
        check_non_decreasing()
       
	   
        for i in range(k, len(nums) - k):
            if  non_increasing_dp[i - 1] >= k and  non_decreasing_dp[i + 1] >= k:
                ans.append(i)
        return ans 
